### ISSUE TITLE
Color with A_BOLD "different" between WinGUI/X11

### ISSUE THREAD

**Issue by rmorales87atx** [Bill-Gray/PDCursesMod#178]
Hello,

I have my colors set up using this code: https://pastebin.com/LNtNdSj8

What I've noticed is when I have a "background color" set on something with an `A_BOLD` attribute, the result isn't quite what I expect on WinGUI vs X11. 

Here's an example from my actual application: this menu should have its selected item as bright white on the red background.

|WinGUI|X11|
|-|-|
|![void_menu.png](http://squailboont.info/img/void_menu.png)|![void_menu_x11.png](http://squailboont.info/img/void_menu_x11.png)|

I pulled the latest code as of this post, so my "version" of PDCurses is `4.1.99`. It's probably good to note that this app calls `PDC_set_bold(false)`, since I prefer having bold = bright instead of a bold font-face. I haven't yet tested any other flavors besides WinGUI and X11. (In a much earlier prototype of this app, the Linux version actually linked against ncurses; I can't recall if ncurses displayed the menus like X11 did.)

Perhaps this is more likely a "bug" in my understanding of colors & attributes, and not a bug with PDCurses. Can anyone explain why these menus look different on WinGUI & X11?

**Comment by Bill-Gray** [comment]
Hmmm... perhaps I'm not actually understanding the problem,  but I think you're just seeing the fact that WinGUI defaults to use of a [Courier New font](https://github.com/Bill-Gray/PDCursesMod/blob/126a8df8f33e64439a12e7ecbe0685804d0ef00c/wingui/pdcdisp.c#L207),  and X11 to the [misc-fixed fonts](https://github.com/Bill-Gray/PDCursesMod/blob/126a8df8f33e64439a12e7ecbe0685804d0ef00c/x11/pdcscrn.c#L20).  WinGUI makes it reasonably easy for a user to choose a font;  X11,  at least at present,  doesn't.  A bit of tweaking of the indicated lines can get you a different font for either of them.

This would all be independent of the background color or `A_BOLD` attribute.  The platforms just use different fonts by default... does this help,  or is it an answer to a problem you don't actually have?

**Comment by rmorales87atx** [comment]
The issue isn't the fonts. I'll try to better describe in words, maybe you aren't able to see the embedded images. The pictured menu displays the current selected item with a red on white color pair combined with the bold attribute so that the white background is a bright white instead of a dim white. On WinGUI the bright white does not show, instead it shows a dim red on dim white. On X11 the red on bright white appears as I expect.

**Comment by Bill-Gray** [comment]
Sorry for the late reply!

I _was_ misunderstanding the problem,  even with the images.  Now that I've gotten what's going on into my head,  I think the trouble is at [this line in `pdccolor.c'](https://github.com/Bill-Gray/PDCursesMod/blob/126a8df8f33e64439a12e7ecbe0685804d0ef00c/common/pdccolor.c#L197) (and following lines).

You'll see that if the text is reversed,  we swap foreground and background color.  Then,  if the text is in bold,  but is to be highlit rather than shown in a for-real bold font,  we intensify the foreground color.

In X11,  we'd first intensify the foreground color,  and _then_ do the swap.  I guess I'd lean toward doing things that way,  so that 'reversed colors' are truly reversed.  In other words,  that `if( reverse_colors)` block should be the last thing we do.

So I'd suggest [replacing `common/pdccolor.c` with this code](https://www.projectpluto.com/temp/pdccolor.c) and seeing if that does the trick.

**Comment by GitMensch** [comment]
Hm, that fix would also apply to VT, so it may be reasonable to adjust the title.
@rmorales87atx Can you please make a new screenshot after applying the patch to WinGUI?
@Bill-Gray Do we have a test (demo) to easy compare the non-blinking and non-bold (=brightened) results with its attributes?

**Comment by rmorales87atx** [comment]
I'm happy to report that it looks like you got it on the first go! 😀

![void_menu](https://user-images.githubusercontent.com/3020857/90237754-684a3180-ddea-11ea-9349-e5da87608390.png)
![sales_reduction](https://user-images.githubusercontent.com/3020857/90237773-6da77c00-ddea-11ea-9147-0dbce61742c8.png)

**Comment by Bill-Gray** [comment]
@rmorales87atx - excellent.  I've pushed that change.
@GitMensch - yes,  this applies at present to WinGUI and VT.  It will eventually apply to X11,  the SDLs,  and possibly WinCon and DOSVGA : i.e.,  any platform that allows a full range of RGBs really should be using this code,  thereby eliminating redundancy and extending full color support to those platforms (they currently have limited palettes).  The logic really should be the same on all non-palette-based platforms.

This is not a priority item,  though,  at least not for me.  I'd rather like to have full color support on X11,  so that's the one I'm apt to fix first.  Actually looks as if it would be fairly straightforward to do...

The `newtest` demo shows this effect.  Click on 'Bold On' or 'Bold Off',  and you can see the "Standout text" go from the use of a for-real bold font to instead showing a standard font with an intensified foreground.  (Note that in X11,  this switch doesn't take effect until you resize the window.  The other platforms have code to redraw all bold text when `PDC_set_bold()` is called and changes the "boldness" state.  X11,  at present,  does not.  But if you do resize the window,  you'll see the change.)

**Comment by GitMensch** [comment]
>  X11, at present, does not. But if you do resize the window, you'll see the change.

I'd say that's a bug to be solved :-(

**Comment by Bill-Gray** [comment]
> I'd say that's a bug to be solved :-(

More generally,  the need to resize the screen,  or otherwise force the text to be redrawn,  is present on all platforms for wmcbrine's PDCurses and in all platforms of PDCursesMod except for WinGUI and VT.  Similarly,  calls to `PDC_init_pair()` don't take effect until the text using that color pair is redrawn,  except on palette-based platforms such as DOS and OS/2.

**Comment by GitMensch** [comment]
Hm, could we internally change the size of the window for wincon and x11 to "1 line more", then "back to 1 line less" actually forcing a redraw?

**Comment by Bill-Gray** [comment]
Many,  if not most, actions in PDCurses/ncurses don't take effect until `refresh()` is called.  Put text on the screen,  reset a palette entry or color pair,  and nothing happens right away;  it's all saved up for the moment `refresh()` is called.  It seems reasonable to me for `PDC_set_bold()` and `PDC_set_blink()` to behave similarly.  It's fine if they take effect immediately,  as they do on some platforms.  But it's probably best that they do nothing (except mark bold or blinking text as 'dirty') until the next time `refresh()` is called.

Similarly for calls to `init_pair()` and `init_color()` : on many platforms,  including ncurses,  neither function does anything until `refresh()` is called.  The exception,  which is hard to avoid,  is that `init_color()` will result in an immediate response on systems with hardware color palettes (DOS,  OS/2,  DOSVGA in 8-bit and 4-bit palette modes). 

I've written [a small test program](https://www.projectpluto.com/temp/boldtest.c) to investigate the behavior.  When compiled with ncurses,  nothing changes until `refresh()` is called.  Similarly for some PDCurses platforms.  VT responds to such changes immediately.  WinGUI responds to color changes immediately,  but to bold/blink changes within 0.5 seconds (waits to the next 'blink screen' command.) 

But there are some oddities.  The `SP->dirty` flag is set when you call `init_color()`,  causing a redraw when `napms()` is called.  So you _might_ get a "palette" update on a non-hardware-palette system,  but not until (usually) `getch()` is called.  You _can_ get situations where a palette entry is changed and text using that palette entry isn't updated.  I think I have some fixing to do...

All of this,  by the way,  applies to wmcbrine's version as well.

**Comment by Bill-Gray** [comment]
After a fair bit of investigation,  I realized the sensible thing to do here is : when `init_color()`,  `init_pair()`,  `PDC_set_blink()`,  or `PDC_set_bold()` are called,  we set `curscr->_clear` to `TRUE`;  i.e.,  we tell PDCurses,  "Next time you update the screen,  update absolutely everything."

Ideally,  we might do a more measured approach;  if you init color pair 3,  we'd only redraw the text with that pair,  for example.  But that turns out to be more difficult than I thought;  PDCurses is somewhat bright about checking for changes,  and would require considerable additional logic to note changes of this ilk.  And I can think of zero actual current cases where it would matter... we'd be looking at a good bit of extra code for very little benefit.
